### DISCUSSION TITLE
X axis alignment issues

### DISCUSSION THREAD

**Discussion by jdeltoft** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#547] [Q&A]
Is anyone else chaining displays and seeing any x offset issues?  I have 4 panels (64x64) chained in an overall square, and I get alignment issues from the top two to the bottom two for example.  It all starts with the very first test display image.  I added one +1 to this code below as "`top_left_x  + 1`" , which mostly fixed it but the very final edge is still missing.  

```
inline void VirtualMatrixPanel::drawDisplayTest()
{   
  // Write to the underlying panels only via the dma_display instance.
    this->display->setFont(&FreeSansBold12pt7b);
    this->display->setTextColor(this->display->color565(255, 255, 0));
    this->display->setTextSize(1);
    
    for (int panel = 0; panel < vmodule_cols * vmodule_rows; panel++)
    {
        int top_left_x = (panel == 0) ? 0 : (panel * panelResX);
        this->display->drawRect(top_left_x  + 1  , 0, panelResX, panelResY, this->display->color565(0, 255, 0));
        this->display->setCursor((panel * panelResX) + 2, panelResY - 4);
        this->display->print((vmodule_cols * vmodule_rows) - panel);
    }
}   
```

Here is the picture of this:

Before:
<img width="639" alt="image" src="https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/4506303/112a7e82-a3f2-4a2f-9bd0-ec4e3454f066">

After:
<img width="628" alt="image" src="https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/4506303/ad2056ed-7a65-4749-a471-ceaab99ef2db">

But as you see, in the AFTER photo, there is a column missing still on the right side of the `1` panel.

I need to now go translate this "fix" to the virtual display since my album art is also experiencing this. It's harder to tell the exact offset but it's probably just one still.

<img width="628" alt="image" src="https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/4506303/13787453-05e2-4fba-b54d-8ba50008e07c">

**Comment by mrcodetastic** [comment]
Clkphase value?

**Comment by jdeltoft** [comment]
Let me check on that value.  I did just "fix" the album art with this obvious change per the above.  

```
inline void VirtualMatrixPanel::drawPixelRGB888(int16_t x, int16_t y, uint8_t r, uint8_t g, uint8_t b)
{   
    this->getCoords(x, y);
    this->display->drawPixelRGB888(coords.x+1, coords.y, r, g, b);
}   
```

**Comment by jdeltoft** [comment]
Hmmm, apologies, but I don't see any setting for clock phase, so I guess default?

**Comment by jdeltoft** [comment]
Interesting, I see the green boarder on the far left side not being overwritten?  Odd.  The missing column on the top right is still there also.  But you can see the letters are perfectly aligned now between the vertical halfway point.

<img width="633" alt="image" src="https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/4506303/4be73f52-dbcb-4fc7-84b9-17bd2df2f16b">

**Comment by jdeltoft** [comment]
Ignore that comment, that was me doing something in the main code. Just the final column issue remains

**Comment by jdeltoft** [comment]
It's telling that the test function also fails per comment#1 above, but the other green outline squares are perfect.  So it's some limit for the total display writable area? I could try reducing the panel count to 1 and see if that fails for the one side of a boarder square I guess.

**Comment by jdeltoft** [comment]
Fixed this, just a simple change for the offset.  I will post a patch later
